### ISSUE TITLE
CA1305 doesn't catch implicit ToString calls, e.g. StringBuilder.Append(object)

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn-analyzers#4518]
### Analyzer

**Diagnostic ID**: [CA1305](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1305`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

There should be a warning that ToString is implicitly called without specifying IFormatProvider in these cases, but there is not:

```cs
using System;
using System.IO;
using System.Text;

class C
{
    void M(int p)
    {
        // There should be a warning for each line (not an exhaustive list):

        new StringBuilder().Append(p);

        ((TextWriter)new StringWriter()).Write(p);
        ((TextWriter)new StringWriter()).WriteLine(p);

        Console.Write(p);
        Console.WriteLine(p);

        _ = string.Join(", ", new[] { p });

        _ = p + "";

        _ = $"{p}";
    }
}
```

Example of a place where it is correct to have no warning:

```cs
using System;

class C
{
    void M(int p)
    {
        // Should not begin warning:
        _ = (FormattableString)$"{p}";
    }
}
```

### Evidence that missing implicit this is bad

Super subtle compiler bug due to the compilation server spinning up with a `sv-SE` culture:
https://github.com/dotnet/roslyn/issues/49826#issuecomment-739813166

**Comment by mavasani** [comment]
We can do couple of things here:
1. Extend the analyzer to also analyze StringBuilder.Append method
2. Expose an editorconfig option to allow custom user-defined methods to be included in analysis.

Tagging @Evangelink who has helped in implement similar features in the past.

**Comment by jnm2** [comment]
@mavasani Would you consider building in knowledge of TextWriter.Write/Line and Console.Write/Line as well as StringBuilder.Append?

**Comment by mavasani** [comment]
@jnm2 Yes, it would be fine to special case and analyze by default any BCL methods which fall in this bucket.

**Comment by Youssef1313** [comment]
Are there any performance implications of encouraging `builder.Append(myInt.ToString(formatProvider))`?
I think `builder.Append(myInt)` works on spans and might be less allocating than calling `ToString`.
